File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.py
from collections import deque

N, K = map(int, input().split())
visited = [-1]*100001

def bfs(now, target):
    q = deque([now])
    visited[now] = 0

    while q:
        x = q.popleft()
        next_pos = [x+1, x-1, x*2]
        for np in next_pos:
            if 0<=np<100001 and visited[np]==-1:
                q.append(np)
                visited[np] = visited[x] + 1
    return visited[target]


result = bfs(N, K)

# print(visited[:50])
lst = deque([K])
while True:
    # print(lst)
    if lst[0]%2==0:
        next_pos = [lst[0]+1, lst[0]-1, lst[0]//2]
    else:
        next_pos = [lst[0]+1, lst[0]-1]
    # print(next_pos)
    for np in next_pos:
        # print()
        # print('lst[0]: ', lst[0])
        # print('np: ', np)
        # print()
        if 0<=np<100001 and visited[lst[0]]-1 == visited[np]:
            lst.appendleft(np)
            break
    if lst[0]==N:
        break

print(result)
print(*lst)

File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.cc
#include <iostream>
#include <queue>
#include <vector>
#include <tuple>
#include <deque>

#define MAX 1000001

using namespace std;

int N, K;
vector<int> visited;
deque<int> route;

int bfs(){
    queue<int> q;
    q.push(N);
    visited[N] = 0;

    while(!q.empty()){
        int now = q.front();
        q.pop();

        for(int nx : {now-1, now+1, now*2}){
            if(0<=nx && nx<MAX && visited[nx]==-1){
                q.push(nx);
                visited[nx] = visited[now] + 1;
            }
        }
    }

    int now = K;
    while(1){
        route.push_front(now);
        if(now==N){
            break;
        }
        if(now%2==0 && visited[now/2]==visited[now]-1){
            now /= 2;
        }
        else{
            for(int nx : {now-1, now+1}){
                if(0<=nx && nx<MAX && visited[nx]==visited[now]-1){
                    now = nx;
                    break;
                }
            }
        }
    }
    return visited[K];
}

int main(void){
    // freopen("input.txt", "r", stdin);

    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> N >> K;
    visited = vector<int>(MAX, -1);

    int result = bfs();

    cout << result << '\n';
    for(int r : route){
        cout << r << ' ';
    }

    return 0;
}